pkg/controller/atlasprivateendpoint/privateendpoint.go

```diff
@@ -0,0 +1,236 @@
+		}
+	}
+
+	switch atlasPEService.Status() {
+	case privateendpoint.StatusInitiating:
+		return r.inProgress(ctx, akoPrivateEndpoint, atlasPEService, api.PrivateEndpointServiceReady, workflow.PrivateEndpointServiceInitializing, "Private Endpoint is being initialized")
+	case privateendpoint.StatusPending, privateendpoint.StatusPendingAcceptance, privateendpoint.StatusWaitingForUser, privateendpoint.StatusVerified:
+		return r.inProgress(ctx, akoPrivateEndpoint, atlasPEService, api.PrivateEndpointServiceReady, workflow.PrivateEndpointServiceInitializing, "Private Endpoint is waiting for human action")
+	case privateendpoint.StatusFailed, privateendpoint.StatusRejected:
+		return r.terminate(ctx, akoPrivateEndpoint, atlasPEService, api.PrivateEndpointServiceReady, workflow.PrivateEndpointServiceFailedToConfigure, errors.New(atlasPEService.ErrorMessage()))
+	case privateendpoint.StatusDeleting:
+		return r.inProgress(ctx, akoPrivateEndpoint, atlasPEService, api.PrivateEndpointServiceReady, workflow.PrivateEndpointServiceDeleting, "Private Endpoint is being deleted")
+	}
+
+	ctx.SetConditionTrue(api.PrivateEndpointServiceReady)
+	r.EventRecorder.Event(akoPrivateEndpoint, "Normal", string(workflow.PrivateEndpointServiceCreated), "Private Endpoint Service is available")
+
+	return r.handlePrivateEndpointInterface(ctx, projectID, akoPrivateEndpoint, akoPEService, atlasPEService)
+}
+
+func (r *AtlasPrivateEndpointReconciler) handlePrivateEndpointInterface(
+	ctx *workflow.Context,
+	projectID string,
+	akoPrivateEndpoint *akov2.AtlasPrivateEndpoint,
+	akoPEService privateendpoint.EndpointService,
+	atlasPEService privateendpoint.EndpointService,
+) (ctrl.Result, error) {
+	compositeInterfacesMap := privateendpoint.MapPrivateEndpoints(akoPEService.EndpointInterfaces(), atlasPEService.EndpointInterfaces())
+
+	if len(compositeInterfacesMap) == 0 {
+		return r.waitForConfiguration(ctx, akoPrivateEndpoint, atlasPEService)
+	}
+
+	// we want to sync all interface, if any of them is in progress, after all we transition to in progress
+	pendingResources := false
+	for _, compositeInterfaceMap := range compositeInterfacesMap {
+		inAKO := compositeInterfaceMap.AKO != nil
+		inAtlas := compositeInterfaceMap.Atlas != nil
+		inProgress := isInterfaceInProgress(compositeInterfaceMap.Atlas)
+		failed := hasInterfaceFailed(compositeInterfaceMap.Atlas)
+
+		switch {
+		case failed:
+			return r.terminate(
+				ctx,
+				akoPrivateEndpoint,
+				atlasPEService,
+				api.PrivateEndpointReady,
+				workflow.PrivateEndpointFailedToConfigure,
+				errors.New(compositeInterfaceMap.Atlas.ErrorMessage()),
+			)
+		case inProgress:
+			pendingResources = true
+			continue
+		case inAKO && !inAtlas:
+			gcpProjectID := getGCPProjectID(akoPrivateEndpoint, compositeInterfaceMap.AKO.InterfaceID())
+			_, err := r.privateEndpointService.CreatePrivateEndpointInterface(
```

Comment:
Sorry, but I also didn't get what you mean here. Could you improve the example?

Reply:
You could call a createPEI method (name is just an example) which is already multi-provider aware so we do not leak the GCP specific query at this method, but hide it within that method. I think it would make more sense that way, for readability as that call works at a lower abstraction level than this logic here.